src/rgw/cls_fifo_legacy.cc

```diff
@@ -440,7 +440,13 @@ int FIFO::apply_update(const DoutPrefixProvider *dpp,
   ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__ << ":" << __LINE__
 		 << " entering: tid=" << tid << dendl;
   std::unique_lock l(m);
-  if (objv != info->version) {
+
+  if (objv < info->version) {
```

Comment:
This is the one part I'm concerned about, since if we have a /newer/ version than is being expected that also indicates we've got out of sync.

Reply:
Yes it is a out-of-sync situation.  Here is what I'm thinkging.
When the code reaches here, it means 
1) the update_meta succeeded, witch in turn means, in this case, the journal entries have been added in;
2) a newer version has been retrieved by another thread, and the newly added journal entries must be in this newer version (otherwise the above item 1 wouldn't have succeeded).   
So in this case, we have to to give the rgw a chance to remove the journal entries from the remote, which is why I return 0  here.

Reply:
The earlier successful _update_meta has already written the journal entries to the backend, and all the later versions will carry them before they are removed by somebody. And they also block further update on this part_num. The problem now is that nobody is deleting them. If we return non-zero here and force a re-read, no deletion is triggered.
  
Example in the race condition 2 in [the tracker 5762](https://tracker.ceph.com/issues/57562).
"journal: {{{2=op: Op::set_head, part_num: 2, part_tag: }}" has been left in the datalog backend meta data.

Some log lines: 
2022-11-12T11:52:29.472-0500 7f3b4bf4f700  0 req 14970605871587688978 0.042999800s s3:put_obj int rgw::cls::fifo::FIFO::_prepare_new_head(const DoutPrefixProvider*, int64_t, uint64_t, optional_yield):870 tid=1779 new_head_part_num=2 info=(id: data_log.13, version: 4OkaLaZsVoIVzNgB{7}, oid_prefix: data_log.13.5Ix6oXD/TeT, params: {max_part_size: 1048576, max_entry_size: 32768, full_size_threshold: 1015750}, tail_part_num: 0, head_part_num: 2, min_push_part_num: 0, max_push_part_num: 2, head_tag: 7LedASIbULCce4HS, tags: {{2=7LedASIbULCce4HS}}, journal: {{{}})
2022-11-12T11:52:29.472-0500 7f3b4bf4f700  0 req 14970605871587688978 0.042999800s s3:put_obj int rgw::cls::fifo::FIFO::_prepare_new_head(const DoutPrefixProvider*, int64_t, uint64_t, optional_yield):905 tid=1779 jentry=(op: Op::set_head, part_num: 2, part_tag: )
...
2022-11-12T11:52:29.492-0500 7f3b4bf4f700  0 req 14970605871587688978 0.062999710s s3:put_obj int rgw::cls::fifo::FIFO::apply_update(const DoutPrefixProvider*, rados::cls::fifo::info*, const rados::cls::fifo::objv&, const rados::cls::fifo::update&, uint64_t):442 tid=1779 objv=4OkaLaZsVoIVzNgB{7} *info=(id: data_log.13, version: 4OkaLaZsVoIVzNgB{8}, oid_prefix: data_log.13.5Ix6oXD/TeT, params: {max_part_size: 1048576, max_entry_size: 32768, full_size_threshold: 1015750}, tail_part_num: 0, head_part_num: 2, min_push_part_num: 0, max_push_part_num: 2, head_tag: 7LedASIbULCce4HS, tags: {{2=7LedASIbULCce4HS}}, journal: {{{2=op: Op::set_head, part_num: 2, part_tag: }})

When we reach the apply_update, we can see the info has a higher version ("4OkaLaZsVoIVzNgB{8}")  than this operation ("4OkaLaZsVoIVzNgB{7}"). And in this version "4OkaLaZsVoIVzNgB{8}",  head_part_num and max_push_part_num have already been updated to 2. 

According to the following code line, the _prepare__new_part will return succeess from here, and no one will handle the left behind journal entry "journal: {{{2=op: Op::set_head, part_num: 2, part_tag: }}".
https://github.com/adamemerson/ceph/blob/wip-57562/src/rgw/cls_fifo_legacy.cc#L789

Reply:
hmm.. never mind, it looks like returning 0 has the same effect. 
I guess this code block should be able to fix both race conditions.
https://github.com/adamemerson/ceph/pull/1/files#diff-bc606c53e2823b90986c53aab0cb69b3ef23f1e19f5b8fa7b29eb70889a806e4R393

Let me clean up this PR. Some of the code is not necessary. Will test with the clean up version  as well.